src/Altinn.App.Core/Implementation/DefaultTaskEvents.cs

```diff
@@ -160,20 +160,77 @@ public async Task OnEndProcessTask(string endEvent, Instance instance)
+        var dataGuid = Guid.Parse(dataElement.Id);
+        string app = instance.AppId.Split("/")[1];
+        int instanceOwnerPartyId = int.Parse(instance.InstanceOwner.PartyId);
+        object data = await _dataClient.GetFormData(
+            instanceGuid, modelType, instance.Org, app, instanceOwnerPartyId, dataGuid);
+
+        // Remove hidden data before validation, ignore hidden rows. TODO: Determine how hidden rows should be handled going forward.
         if (_appSettings?.RemoveHiddenData == true)
         {
-            await RemoveHiddenData(instance, instanceGuid, dataTypesToLock);
+            var layoutSet = _appResources.GetLayoutSetForTask(dataType.TaskId);
+            var evaluationState = await _layoutEvaluatorStateInitializer.Init(instance, data, layoutSet?.Id);
+            LayoutEvaluator.RemoveHiddenData(evaluationState, RowRemovalOption.Ignore);
         }
-    }
 
-    private async Task RunRemoveShadowFields(Instance instance, Guid instanceGuid, List<DataType> dataTypesToLock)
-    {
-        if (dataTypesToLock.Find(dt => dt.AppLogic?.ShadowFields?.Prefix != null) != null)
+        // Remove shadow fields
+        if (dataType.AppLogic?.ShadowFields?.Prefix != null)
         {
-            await RemoveShadowFields(instance, instanceGuid, dataTypesToLock);
+            var modifier = new IgnorePropertiesWithPrefix(dataType.AppLogic.ShadowFields.Prefix);
+            JsonSerializerOptions options = new()
+            {
+                TypeInfoResolver = new DefaultJsonTypeInfoResolver
+                {
+                    Modifiers = { modifier.ModifyPrefixInfo }
+                },
+                DefaultIgnoreCondition = JsonIgnoreCondition.WhenWritingNull,
+            };
+
+            string serializedData = JsonSerializer.Serialize(data, options);
+            if (dataType.AppLogic.ShadowFields.SaveToDataType != null)
+            {
+                // Save the shadow fields to another data type
+                var saveToDataType = dataTypesToLock.Find(dt => dt.Id == dataType.AppLogic.ShadowFields.SaveToDataType);
+                if (saveToDataType == null)
+                {
+                    throw new Exception($"SaveToDataType {dataType.AppLogic.ShadowFields.SaveToDataType} not found");
+                }
+
+                Type saveToModelType = _appModel.GetModelType(saveToDataType.AppLogic.ClassRef);
+                var updatedData = JsonSerializer.Deserialize(serializedData, saveToModelType);
+                await _dataClient.InsertFormData(updatedData, instanceGuid, saveToModelType ?? modelType, instance.Org, app, instanceOwnerPartyId, saveToDataType.Id);
+            }
+            else
+            {
+                // Remove the shadow fields from the data
+                data = JsonSerializer.Deserialize(serializedData, modelType)!;
+            }
         }
+        // remove AltinnRowIds
+        ObjectUtils.RemoveAltinnRowId(data);
```

Comment:
Should we always remove the row id upon completion?

Reply:
Yup, I believe so. They're strictly only needed for us, internally when filling out a form. If app developers have other needs - such as requiring that rows are unique in some way, where no rows may contain the same data as any other row - that is never handled by our row ID anyway, and would have to be implemented using validation messages.